src/resolvers.ts

```diff
@@ -0,0 +1,130 @@
+    ): Book => {
+      const newBook = new Book(title, author);
+      booksDataSource.addBook(newBook);
+      return newBook;
+    },
+
+    // Add a new customer
+    addCustomer: (
+      _: any,
+      {name, email }: {name: string; email: string }
+    ): Customer => {
+      const newCustomer = new Customer(name, email);
+      customersDataSource.addCustomer(newCustomer);
+      return newCustomer;
+    },
+
+    // Lend a book to a customer, updates the fields of a book related to lending
+    lendBook: (
+      _: any,
+      {
+        bookId,
+        customerId,
+        lentDate,
+        dueDate,
+      }: {
+        bookId: string;
+        customerId: string;
+        lentDate: string;
+        dueDate: string;
+      }
+    ): Book => {
+      const book = booksDataSource.getBookById(bookId);
+      const customer = customersDataSource.getCustomerById(customerId);
+
+      if (book && customer && !book.isLent) {
+        const record = new LendingRecord(customer.id, lentDate, dueDate);
+
+        book.isLent = true;
+        book.currentLendeeId = customer.id;
+        book.lendingHistory.push(record);
+        booksDataSource.updateBook(book);
+
+        return book;
+      } else {
+        const errorMessage = `Book with ID ${bookId} is already lent or does not exist.`;
+        logger.error(errorMessage);
+        throw new Error(errorMessage);
+      }
+    },
+
+    // Mark a book as returned, change its isLent field to false and add a new lendingRecord
+    returnBook: (
+      _: any,
+      { bookId, returnDate }: { bookId: string; returnDate: string }
+    ): Book => {
+      const book = booksDataSource.getBookById(bookId);
+      // Check if the book exists and is actually lent to someone
+      if (book && book.isLent) {
+        // Find the lending record without return date
+        const record = book.lendingHistory.find((r) => !r.returnDate);
+        // Update the record and the book
+        if (record) {
+          record.returnDate = returnDate;
+          book.isLent = false;
+          book.currentLendeeId = undefined;
+          booksDataSource.updateBook(book);
+
+          return book;
+        }
+      }
+      const errorMessage = `Book with ID ${bookId} is not currently lent or does not exist.`;
+      logger.error(errorMessage);
+      throw new Error(errorMessage);
+    },
+  },
+
+  Book: {
+    // Find the current customer that is lending the book
+    currentLendee: (book: Book): Customer | null => {
+      if (book.currentLendeeId)
+        return customersDataSource.getCustomerById(book.currentLendeeId)!;
+      return null;
+    },
+
+    // Find the lending history
+    lendingHistory: (book: Book): LendingRecord[] => {
+      return book.lendingHistory;
+    },
+  },
+
+  LendingRecord: {
+    // Find the customer associated with a lending record
+    customer: (record: LendingRecord): Customer => {
+      return customersDataSource.getCustomerById(record.customerId)!;
```

Comment:
Hmmm. Something smells fishy here. Why do you have to use the assertion operator here?

Reply:
I used it because otherwise the return type would have been `Customer | undefined`, but if we have a lending record the only way to already have it is to have called the `lendBook` method, and it already has assigned valid customer to the lending record. So if I follow the logic if we have a lending record, we already have the customer in the database, however typescript can't be sure about it and therefore I ma using the assertion operator.